Запретить реакт-вью непосредственно внутри яте-бокса

vitkarpov opened the pull request.

vitkarpov committed:
Тест #41

vitkarpov commented:
@Rebulus

vitkarpov commented on test/integration/ns.viewReact.js:
```diff
@@ -859,28 +862,38 @@ describe('ns.ViewReact интеграционные тесты ->', function() {
                     .render()
                     .then(function() {
-                        expect($('body').find(this.nodeV1)).to.have.length.above(0);
+                        expect(this.isViewInDOMOnComponentDidMount).to.be.true;
```
Это нечестная проверка, потому что на момент вызова этого кода нода всегда будет в DOM (просто ссылка в this.nodeV1 была записана на componentDidMount). А нужен именно факт наличия на момент вызова колбека.

Rebulus commented:
осталось понять, что идёт не так  😄

Rebulus commented:
@mishk0, @vitkarpov Подебажил, да, дело в `_sortViewNodes`. На второй шаг создаётся новая вьюшка для `app3` лайаута,  т.к. был изменён ключ. Она рендрится и в итоге создание компонента происходит до того, как бокс заменит предыдущее реактивное view на новое.

vitkarpov commented:
Кстати, а почему при первой отрисовке такого нет? Вроде `_sortViewNodes` всегда вызывается после `_updateHTML`, как написал @mishk0

Rebulus commented:
При первой отрисовке мы имеем полное дерево отображения, а не единичную вьюшку. В итоге, т.к. `app` корневой, то он вставит сразу весь сет в DOM, а лишь потом пойдёт по деткам. Где и случится в итоге рендер компонента.

vitkarpov commented:
Окай :)

vitkarpov commented:
Остально понять, как это правильно б зачинить :)

Rebulus commented:
https://github.com/yandex-ui/noscript/blob/master/src/ns.view.js#L1256

Rebulus commented:
https://github.com/yandex-ui/noscript/blob/master/src/ns.view.js#L1286-L1288

Rebulus commented:
В `ns` на данный момент нет инструмента понимания, что бокс переставляет вьюшки. Он прячет эту логику в себе. Моё предложение, пока запретить использование обычных боксов в сочетании с реактивной вьюшкой в теле).

vitkarpov commented:
В принципе, всегда можно воткнуть реактивную вьюшку и положить в нее бокс.

Rebulus commented:
Если потребуется сохранить `yate` бокс, то в него надо будет ложить `yate` вью, а вот в ней уже можно `reactView`.

Rebulus commented:
Но надо подумать ещё чутка, вдруг придёт решение 😄

Rebulus commented:
Как вариант, можно ещё вынести `_renderElement` на `ns-view-htmlinit` файзу.

Rebulus commented:
@vitkarpov может быть есть какие-то новости про это ишью?

vitkarpov commented:
Пока все так же: не придумали как чинить, и дергаем `_renderElement` на `ns-view-htmlinit` в той вьюшке, где это актуально.

Rebulus commented:
Понятно, жаль(

vitkarpov commented:
Мне кажется, что надо это просто закрыть и в качестве фикса «запретить использование обычных боксов в сочетании с реактивной вьюшкой в теле» 🐲 

Кажется, что с точки зрения приложения всегда можно этого избежать — надо только в документации отметить. Потому что задачи поддержать все возможные взаимодействия ns+react нет — надо дать возможность писать часть дерева на реакте.

vitkarpov commented:
Можно кидать `console.warning` при попытке создать реактовую вьюху внутри яте-бокса. Для разработчика будет понятно что происходит.

vitkarpov committed:
[#62] Запретить использование реак-вью непосредственно внутри яте-бокса

- кидаем исключение с понятным сообщением (для удобного переезда)
- правим тесты (добавляем везде реакт-обертку над боксом, где актуально)

vitkarpov committed:
eslint

Rebulus commented:
Согласен. Можно вернуться, при необходимости, к этому ишью. Но пока, более простой способ - это кинуть исключение, как создание вьюхи с неопределённого конструктора и т.п.
Давай закрывать 😄
